proposed/activity-improvements.md

```diff
@@ -0,0 +1,465 @@
+        public ActivitySource(string name, string version = "");
+
+        /// Returns the ActivitySource name.
+        public string Name { get; }
+
+        /// Returns the ActivitySource version.
+        public string Version { get; }
+
+        /// Check if there is any listeners for this ActivitySource.
+        /// This property can be helpful to tell if there is no listener, then no need to create Activity object
+        /// and avoid creating the objects needed to create Activity (e.g. ActivityContext)
+        /// Example of that is http scenario which can avoid reading the context data from the wire.
+        public bool HasListeners();
+
+        /// Creates a new Activity object if there is any listener to the Activity, returns null otherwise.
+        public Activity? StartActivity(string name, ActivityKind kind = ActivityKind.Internal);
+
+        /// Creates a new <see cref="Activity"/> object if there is any listener to the Activity events, returns null otherwise.
+        public Activity? StartActivity(
+                            string name,
+                            ActivityKind kind,
+                            ActivityContext parentContext, 
+                            IEnumerable<KeyValuePair<string, string?>>? tags = null,
+                            IEnumerable<ActivityLink>? links = null,
+                            DateTimeOffset startTime = default);
+
+        /// Creates a new <see cref="Activity"/> object if there is any listener to the Activity events, returns null otherwise.
+        public Activity? StartActivity(
+                            string name,
+                            ActivityKind kind,
+                            string parentId,
+                            IEnumerable<KeyValuePair<string, string?>>? tags = null,
+                            IEnumerable<ActivityLink>? links = null,
+                            DateTimeOffset startTime = default);
+
+        /// Dispose the ActivitySource object and remove the current instance from the global list. empty the listeners list too.
+        public void Dispose();
+
+        /// Add a listener to the Activity starting and stopping events.
+        public static void AddActivityListener(ActivityListener listener) { }
+    }
+```
+
+### Activity
+
+```C#
+namespace System.Diagnostics
+{
+    public partial class Activity : IDisposable // added IDisposable
+    {
+        ....
+        public ActivityKind Kind { get; } = ActivityKind.Internal;
+        public string DisplayName { get; set; }
+        public ActivitySource Source { get; }
+
+        public Activity AddEvent(ActivityEvent e);
+        public IEnumerable<ActivityEvent> Events { get; }
+
+        public IEnumerable<ActivityLink> Links { get; }
+
+        public ActivityContext Context { get; }
+
+        public bool IsAllDataRequested { get; set;}
+
+        public void SetCustomProperty(string propertyName, object? propertyValue);
```

Comment:
Would we have Status on Activity? Or we need to use CustomProperty for that?

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#set-status

Reply:
We decided not to introduce the Status now because the current defined OT Status specs are not generic enough for other scenarios. it is useful only for grpc and I believe @SergeyKanzhelev is going to discuss that on the OT side. SetCustomProperty can be used to store your own version of status but expect this may change in the future if decide to introduce Status in Activity.

Reply:
Okay. I'd like to ensure that for every field on Span on OT spec, we should support it as first class in Activity. Using  CustomProperty, which is likely backed by expensive ConcurrentDictionary would be bad for perf right? 

(This is not a blocker, just wanting to make sure we are open to adding first class support to all Span fields to avoid relying on CustomProp)

Reply:
> Using CustomProperty, which is likely backed by expensive ConcurrentDictionary would be bad for perf right?

It depends on your scenario and how the Status will get used and not necessary using ConcurrentDictionary will have bad perf. If we find it has any perf concern we can try to optimize for that scenario. But, anyway, the Status is under discussion and if we reach a conclusion later to introduce a generic form for the Status we can consider adding it to Activity,